feat: add sveltekit example (#176)
antfu/vite-plugin-pwa: docs/frameworks/sveltekit.md

You should remove all references to [SvelteKit service worker module](https://kit.svelte.dev/docs#modules-$service-worker) <outbound-link /> to disable it on your application.

Since `SvelteKit` uses `SSR / SSG`, we need to add the `ReloadPrompt` component using `dynamic import`. `Vite Plugin PWA` will only register the service worker on build, it is aligned with the current behavior of [SvelteKit service worker module](https://kit.svelte.dev/docs#modules-$service-worker) <outbound-link />.

The best place to include the `ReloadPrompt` is on the main layout of the application:

<details>
  <summary><strong>src/routes/__layout.svelte</strong> code</summary>

```html
<script>
    import { onMount } from 'svelte'
    import { browser, dev } from '$app/env'

    let ReloadPrompt
    onMount(async () => {
      !dev && browser && (ReloadPrompt = (await import('$lib/components/ReloadPrompt.svelte')).default)
    })
</script>

<svelte:head>
    {#if (!dev && browser)}
      <link rel="manifest" href="/_app/manifest.webmanifest">
    {/if}
</svelte:head>

<main>
  <slot />
</main>

{#if ReloadPrompt}
  <svelte:component this={ReloadPrompt} />
{/if}
```
</details>

## SvelteKit Adapters

> The main problem with the current implementation of the service worker module of `SvelteKit` is that you don't have access to the result applied by any adapter you have configured on your application.
> 
> The service worker module of `SvelteKit` will be called before the adapter logic is applied, and so, inside the service worker module, you don't have access to those resources.
> 
> Your application will not work when the user is offline, since the pages will not be included on the service worker precache manifest.

Reviewer: These sentences all seem related and so should probably be part of the same paragraph
```suggestion
The main problem with the current implementation of the service worker module of `SvelteKit` is that you don't have access to the result applied by any adapter you have configured on your application. The service worker module of `SvelteKit` will be called before the adapter logic is applied, and so, inside the service worker module, you don't have access to those resources. Your application will not work when the user is offline, since the pages will not be included on the service worker precache manifest.
```

When using `Vite PWA Plugin` with any `SvelteKit Adapter` you need to provide an additional script to rebuild your `pwa` once `SvelteKit` finish building your application, that is, when the adapter configured finish its job.

> The biggest difference between this plugin and the SvelteKit service worker module is that this plugin does not require integration into the application logic, just configuration.
> 
> You can take a look at [SvelteKit example](https://github.com/antfu/vite-plugin-pwa/tree/main/examples/sveltekit-pwa) <outbound-link /> to configure the additional scripts on your application, it is quite complex since we use it for multiple behaviors with the same codebase.

Reviewer: ```suggestion
The biggest difference between this plugin and the SvelteKit service worker module is that this plugin does not require integration into the application logic - just configuration. You can take a look at [SvelteKit example](https://github.com/antfu/vite-plugin-pwa/tree/main/examples/sveltekit-pwa) <outbound-link /> to configure the additional scripts on your application, it is quite complex since we use it for multiple behaviors with the same codebase.
```

> As an example using [@sveltejs/adapter-static](https://github.com/sveltejs/kit/tree/master/packages/adapter-static) <outbound-link /> with `generateSW` strategy and `Prompt for update` behavior, you will need:

Reviewer: ```suggestion
As an example, when using [@sveltejs/adapter-static](https://github.com/sveltejs/kit/tree/master/packages/adapter-static) <outbound-link /> with `generateSW` strategy and `Prompt for update` behavior, you will need:
```

Author: checking the line height on the paragraph with the outbound link icons...

The main problem with the current implementation of the service worker module of `SvelteKit` is that you don't have access to the result applied by any adapter you have configured on your application. The service worker module of `SvelteKit` will be called before the adapter logic is applied, and so, inside the service worker module, you don't have access to those resources. Your application will not work when the user is offline, since the pages will not be included on the service worker precache manifest.

When using `Vite PWA Plugin` with any `SvelteKit Adapter` you need to provide an additional script to rebuild your `pwa` once `SvelteKit` finish building your application, that is, when the adapter configured finish its job.

The biggest difference between this plugin and the SvelteKit service worker module is that this plugin does not require integration into the application logic - just configuration. You can take a look at [SvelteKit example](https://github.com/antfu/vite-plugin-pwa/tree/main/examples/sveltekit-pwa) <outbound-link /> to configure the additional scripts on your application, it is quite complex since we use it for multiple behaviors with the same codebase.

> ### Workbox manifestTransforms

Reviewer: ```suggestion
### Workbox `manifestTransforms`
```

> Before generating the service worker, `manifestTransforms` entry will allow us to transform the resulting precache manifest entries.

Reviewer: some links to the workbox docs here might help
```suggestion
Before generating the service worker, [the `manifestTransforms` option](https://developers.google.com/web/tools/workbox/reference-docs/latest/module-workbox-build#.ManifestTransform) will allow us to transform the resulting precache [manifest entries](https://developers.google.com/web/tools/workbox/reference-docs/latest/module-workbox-build#.ManifestEntry).
```

> The entries provided on `manifestTransforms` (on `workbox` or `injectManifest` pwa plugin option) entry callback from `workbox-build` node module, will contain all the assets specified on the `srcDir` pwa plugin option with the url and their corresponding revision calculated (hash).

Reviewer: ```suggestion
The entries provided in the `manifestTransforms` option (under `workbox` or `injectManifest`) will contain all the assets specified in the `srcDir` option with the URL and their corresponding revision calculated (e.g. to include the hash).
```

> Since `SvelteKit` uses the router name of the directory (for simple cases) for all the generated pages, you can use `manifestTransforms` to modify the url for all pages generated. You must add the logic to do the right mapping using the corresponding adapter.

Reviewer: ```suggestion
Since `SvelteKit` uses the name of the directory (for simple cases) for all the generated pages, you can use `manifestTransforms` to modify the URL for all pages generated. You must add the logic to do the right mapping for the corresponding adapter.
```

> Before generating the service worker, `manifestTransforms` entry will allow us to transform the resulting precache manifest entries.
> 
> The entries provided in the `manifestTransforms` option (under `workbox` or `injectManifest`) will contain all the assets specified in the `srcDir` option with the URL and their corresponding revision calculated (e.g. to include the hash).
> 
> Since `SvelteKit` uses the name of the directory (for simple cases) for all the generated pages, you can use `manifestTransforms` to modify the URL for all pages generated. You must add the logic to do the right mapping for the corresponding adapter.
> 
> Of course, more complex page directories will force you to modify the `manifestTransforms` entry logic. For an example, see the `pwa-configuration.js` module in the next example using `@sveltejs/adapter-static`.

Reviewer: ```suggestion
We must provide a list of URLs for the service worker to load and precache. We provide these to workbox using the [the `manifestTransforms` option](https://developers.google.com/web/tools/workbox/reference-docs/latest/module-workbox-build#.ManifestTransform) under `workbox` or [`injectManifest`](https://developers.google.com/web/tools/workbox/reference-docs/latest/module-workbox-build#.injectManifest).

The [manifest entries](https://developers.google.com/web/tools/workbox/reference-docs/latest/module-workbox-build#.ManifestEntry) provided via the `manifestTransforms` option will contain all the assets specified in the `srcDir` option with the URL and their corresponding revision calculated (e.g. the hash).

Since `SvelteKit` uses the name of the directory (for simple cases) for all the generated pages, you can use `manifestTransforms` to modify the URL for all pages generated. You must add the logic to do the right mapping for the corresponding adapter.

Of course, more complex page directories will force you to modify the `manifestTransforms` entry logic. For an example, see the `pwa-configuration.js` module in the next example using `@sveltejs/adapter-static`.
```

> Of course, a more complex page directories will force you to modify the `manifestTransforms` entry logic, see the `pwa-configuration.js` module in the next example using `@sveltejs/adapter-static`.

Reviewer: ```suggestion
Of course, more complex page directories will force you to modify the `manifestTransforms` entry logic. For an example, see the `pwa-configuration.js` module in the next example using `@sveltejs/adapter-static`.
```

	name: "<YOUR APP NAME>",
	scope: "/",
	start_url: "/",
	display: "standalone",
	theme_color: "#ffffff",
	background_color: "#ffffff",
	icons: [
		{
			src: "/pwa-192x192.png",
			sizes: "192x192",
			type: "image/png"
		},
		{
			src: "/pwa-512x512.png",
			"sizes": "512x512",
			"type": "image/png"
		},
		{
			src: "/pwa-512x512.png",
			"sizes": "512x512",
			"type": "image/png",
			purpose: 'any maskable'
		}
	]
	},
	workbox: {
		// mode: 'development',
		navigateFallback: '/',
		// vite and SvelteKit are not aligned: pwa plugin will use /\.[a-f0-9]{8}\./ by default: #164 optimize workbox work
		dontCacheBustURLsMatching: /-[a-f0-9]{8}\./,
		globDirectory: './build/',
		globPatterns: ['robots.txt', '**/*.{js,css,html,ico,png,svg,webmanifest}'],
		globIgnores: ['**/sw*', '**/workbox-*'],
		manifestTransforms: [async(entries) => {
			// manifest.webmanifest is added always by pwa plugin, so we remove it.
			// EXCLUDE from the sw precache sw and workbox-*
			const manifest = entries.filter(({ url }) =>
				url !== 'manifest.webmanifest' && url !== 'sw.js' && !url.startsWith('workbox-')
			).map((e) => {
				let url = e.url;
				if (url && url.endsWith('.html')) {

> 					if (url.startsWith('/'))
> 						url = url.slice(1)
> 					
> 					e.url = url === 'index.html' ? '/' : `/${url.substring(0, url.lastIndexOf('/'))}`

Reviewer: let's future-proof this so that it works even after https://github.com/sveltejs/kit/issues/1443
```suggestion
					if (url.startsWith('/')) {
						url = url.slice(1)
					}
					if (url === 'index.html') {
						e.url = url === 'index.html' ? '/'
					} else if (url.endsWith('index.html')) {
						e.url = `/${url.substring(0, url.lastIndexOf('/'))}`
					} else if (url.endsWith('.html')) {
						e.url = `/${url.substring(0, url.length - '.html'.length)}`
					}
```

---
title: SvelteKit | Frameworks
---

# SvelteKit

> For `Type declarations`, `Prompt for update` and `Periodic SW Updates` go to [Svelte](/frameworks/svelte.html) entry.
> 

You should remove all references to [SvelteKit service worker module](https://kit.svelte.dev/docs#modules-$service-worker) <outbound-link />
to disable it on your application.

> Since `SvelteKit` will use `SSR / SSG`, we need to add the `ReloadPrompt` component using `dynamic import`. `Vite Plugin PWA`

Reviewer: I'd be curious for more details here as to why SSR / SSG specifically requires the reload prompt. Wouldn't all sites require the reload prompt?

A slightly more natural way to word this would be:
```suggestion
Since `SvelteKit` uses `SSR / SSG`, we need to add the `ReloadPrompt` component using `dynamic import`. `Vite Plugin PWA`
```

will only register the service worker on build, it is aligned with the current behavior of
[SvelteKit service worker module](https://kit.svelte.dev/docs#modules-$service-worker) <outbound-link />.

The best place to include the `ReloadPrompt` is on the main layout of the application:

<details>
  <summary><strong>src/routes/__layout.svelte</strong> code</summary>

```html
<script>

>   import { onMount } from 'svelte'

Reviewer: Svelte files are indented with tabs by default, so it might make sense to use tabs here to make it easier for people to copy/paste

  import { browser, dev } from '$app/env'

  let ReloadPrompt
  onMount(async () => {
    !dev && browser && (ReloadPrompt = (await import('$lib/components/ReloadPrompt.svelte')).default)
  })
</script>

<svelte:head>
  {#if (!dev && browser)}

>   <link rel="manifest" href="/_app/manifest.webmanifest">

Reviewer: ```suggestion
    <link rel="manifest" href="/_app/manifest.webmanifest">
```

  {/if}
</svelte:head>

<main>
  <slot />
</main>

{#if ReloadPrompt}

> <svelte:component this={ReloadPrompt} />

Reviewer: ```suggestion
  <svelte:component this={ReloadPrompt} />
```

{/if}
```
</details>

## SvelteKit Adapters

> The main problem with the current implementation of the service worker module of `SvelteKit` is that you don't have access

Reviewer: I would recommend against manually wrapping the text. If you add a word anywhere it means you need to update every subsequent line. I think it's much easier to manage as one long line. Text editors can always handle wrapping the text automatically

> The main problem with the current implementation of the service worker module of `SvelteKit` is that you don't have access
> to the result applied by any adapter you have configured on your application.

Reviewer: ```suggestion
The main problem with the current implementation of the service worker module of `SvelteKit` is that you don't have access to the result applied by any adapter you have configured on your application.
```

> The main problem with the current implementation of the service worker module of `SvelteKit` is that you don't have access
> to the result applied by any adapter you have configured on your application.
> 
> The service worker module of `SvelteKit` will be called before the adapter logic is applied, and so, inside the service worker
> module, you don't have access to those resources.
> 
> Your application will not work when the user is offline, since the pages will not be included on the service worker precache
> manifest.
> 
> When using `Vite PWA Plugin` with any `SvelteKit Adapter` you need to provide an additional script to rebuild your `pwa`
> once `SvelteKit` finish building your application, that is, when the adapter configured finish its job.
> 
> The biggest difference between this plugin and the SvelteKit service worker module is that this plugin does not require
> integration into the application logic, just configuration.
> 
> You can take a look at [SvelteKit example](https://github.com/antfu/vite-plugin-pwa/tree/main/examples/sveltekit-pwa) <outbound-link />
> to configure the additional scripts on your application, it is quite complex since we use it for multiple behaviors 
> with the same codebase.
> 
> As an example using [@sveltejs/adapter-static](https://github.com/sveltejs/kit/tree/master/packages/adapter-static) <outbound-link />
> with `generateSW` strategy and `Prompt for update` behavior, you will need:

Reviewer: I would recommend against manually wrapping the text. I'm having a hard time reading this on GitHub given all the line breaks. It also means if you add a word anywhere it means you need to update every subsequent line. I think it's much easier to manage as one long line. Text editors can always handle wrapping the text automatically

```suggestion
The main problem with the current implementation of the service worker module of `SvelteKit` is that you don't have access to the result applied by any adapter you have configured on your application.

The service worker module of `SvelteKit` will be called before the adapter logic is applied, and so, inside the service worker module, you don't have access to those resources.

Your application will not work when the user is offline, since the pages will not be included on the service worker precache manifest.

When using `Vite PWA Plugin` with any `SvelteKit Adapter` you need to provide an additional script to rebuild your `pwa` once `SvelteKit` finish building your application, that is, when the adapter configured finish its job.

The biggest difference between this plugin and the SvelteKit service worker module is that this plugin does not require
integration into the application logic, just configuration.

You can take a look at [SvelteKit example](https://github.com/antfu/vite-plugin-pwa/tree/main/examples/sveltekit-pwa) <outbound-link /> to configure the additional scripts on your application, it is quite complex since we use it for multiple behaviors with the same codebase.

As an example using [@sveltejs/adapter-static](https://github.com/sveltejs/kit/tree/master/packages/adapter-static) <outbound-link /> with `generateSW` strategy and `Prompt for update` behavior, you will need:
```

<details>
<summary>1) add <strong>pwa.js</strong> script</summary>

```js
import { resolveConfig } from 'vite'
import { VitePWA } from 'vite-plugin-pwa';
import { pwaConfiguration } from './pwa-configuration.js';
import { copyFileSync } from 'fs';

const webmanifestDestinations = [
	'./.svelte-kit/output/client/',
	'./build/',
]

const swDestinations = [
	'./build/',
]

> const buildPwa = async() => {

Reviewer: it seems like it'd be nice to do this in the Vite plugin. what does it need access to that prevents that from being possible? I wonder if https://github.com/sveltejs/kit/pull/2580 would be enough? Or would you also need to do things like compress assets in a Vite plugin as well?

Author: Here we have a few problems:
1) base path used by SvelteKit is `/_app/` and not `/`: just comment out `base` and `scope` from the pwa configuration and build the example project
2) previous statement also applies for the service worker manifest `scope` and `base` entries
3) the `api` exposed by this plugin is for this purpose: we have integrated on `vite-ssg`, proposed to be included on `vitepress` but rejected by Evan and I have some branch on my SvelteKit repo with the integration

  // const pwaConfiguration = await buildPwaConfiguration()
  const config = await resolveConfig({ plugins: [VitePWA({ ...pwaConfiguration })] }, 'build', 'production' )

>   // when `vite-plugin-pwa` is presented, use it to regenerate SW after rendering

Reviewer: ```suggestion
  // when `vite-plugin-pwa` is present, use it to regenerate SW after rendering
```

  const pwaPlugin = config.plugins.find(i => i.name === 'vite-plugin-pwa')?.api
  if (pwaPlugin?.generateSW) {
    console.log('Generating PWA...')
    await pwaPlugin.generateSW()
    webmanifestDestinations.forEach(d => {
      copyFileSync('./.svelte-kit/output/client/_app/manifest.webmanifest', `${d}/manifest.webmanifest`)
    })

>     // don't copy workbox, svelte kit will copy it

Reviewer: ```suggestion
    // don't copy workbox, SvelteKit will copy it
```

    swDestinations.forEach(d => {
      copyFileSync('./.svelte-kit/output/client/sw.js', `${d}/sw.js`)
    })

>     console.log('Generated PWA complete')

Reviewer: ```suggestion
    console.log('Generation of PWA complete')
```

  includeManifestIcons: false,
  base: '/',
  scope: '/',
  manifest: {
    short_name: "<YOUR APP SHORT NAME>",
    name: "<YOUR APP NAME>",
    scope: "/",
    start_url: "/",
    display: "standalone",
    theme_color: "#ffffff",
    background_color: "#ffffff",
    icons: [
      {
        src: "/pwa-192x192.png",
        sizes: "192x192",
        type: "image/png"
      },
      {
        src: "/pwa-512x512.png",
        "sizes": "512x512",
        "type": "image/png"
      },
      {
        src: "/pwa-512x512.png",
        "sizes": "512x512",
        "type": "image/png",
        purpose: 'any maskable'
      }
    ]
  },
  workbox: {
    // mode: 'development',
    navigateFallback: '/',
    // vite and sveltekit are not aligned: pwa plugin will use /\.[a-f0-9]{8}\./ by default: #164 optimize workbox work
    dontCacheBustURLsMatching: /-[a-f0-9]{8}\./,
    globDirectory: './build/',
    globPatterns: ['robots.txt', '**/*.{js,css,html,ico,png,svg,webmanifest}'],
    globIgnores: [
      '**/sw*', '**/workbox-*'
    ],
    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
    manifestTransforms: [async(entries) => {

>       // manifest.webmanifest is added always by pwa plugin, so we remove it

Reviewer: is this something that should be requested to be changed in the PWA plugin so that it doesn't have to be handled here? maybe what can't be put in the PWA plugin could be provided as a `defaultSvelteKitTransform` or something so that there's not so much that needs to be copy and pasted into each project?

Author: This is an example using `static adapter`, we just traverse generated html files and chnage the name since SvelteKit use that name for the route name, we just need to modify the url, the hash is generated by workbox: this will allow to work offline.

In this callback, workbox will provide the html files with the hash calculated: change the html file name with the router name.

Reviewer: that's helpful details. maybe include it as a code comment?

Author: Do you mean to include my response on the code? Tmr i will include it

Reviewer: Yes, I thought it might be useful to include something similar not your response

      },
      {
        src: "/images/pwa-512x512.png",
        "sizes": "512x512",
        "type": "image/png"
      },
      {
        src: "/images/pwa-512x512.png",
        "sizes": "512x512",
        "type": "image/png",
        purpose: 'any maskable'
      }
    ]
  },
  workbox: {
    mode: 'development',
    navigateFallback: "/",
    // vite and sveltekit are not aligned: pwa plugin will use /\.[a-f0-9]{8}\./ by default: #164 optimize workbox work
    dontCacheBustURLsMatching: /-[a-f0-9]{8}\./,
    globDirectory: './build/',
    globPatterns: ['robots.txt', '**/*.{js,css,html,ico,png,svg,webmanifest}'],
    globIgnores: [
      '**/sw*', '**/workbox-*'
    ],
    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
    manifestTransforms: [async(entries) => {
      // manifest.webmanifest is added always by pwa plugin, so we remove it
      // EXCLUDE from the sw precache sw and workbox-*
      const manifest = entries.filter(({ url }) =>
        url !== 'manifest.webmanifest' && url !== 'sw.js' && !url.startsWith('workbox-')
      ).map((e) => {
        const url = e.url;
        if (url) {
          if (url.endsWith('.html')) {

>             if (url === 'index.html') {
>               e.url = '/';
>             }
>             else {
>               e.url = `/${url.substring(0, url.lastIndexOf('/'))}`;
>             }

Reviewer: you could make this a bit more concise:
```suggestion
            e.url = url === 'index.html' ? '/' : `/${url.substring(0, url.lastIndexOf('/'))}`;
```

Author: updated
